### ISSUE TITLE
TLJH does not automatically create home directories for new users on Ubuntu Server 18.10

### ISSUE THREAD

**Issue by owah** [jupyterhub/the-littlest-jupyterhub#237]
Once tljh was up and running ([after I changed the installer locally to use a newer node version](https://github.com/jupyterhub/the-littlest-jupyterhub/issues/236)), I couldn't start any notebooks, because the spawner process would time out. I have looked into the logging and came upon the following:

```
sudo journalctl -u jupyter-testuser:

Nov 22 12:09:20 blockchainlab systemd[1]: Started /bin/bash -c cd /home/jupyter-testuser && exec jupyterhub-singleuser --port=36531.
Nov 22 12:09:20 blockchainlab systemd[20897]: jupyter-testuser.service: Changing to the requested working directory failed: No such file or directory
Nov 22 12:09:20 blockchainlab systemd[20897]: jupyter-testuser.service: Failed at step CHDIR spawning /bin/bash: No such file or directory
Nov 22 12:09:20 blockchainlab systemd[1]: jupyter-testuser.service: Main process exited, code=exited, status=200/CHDIR
Nov 22 12:09:20 blockchainlab systemd[1]: jupyter-testuser.service: Failed with result 'exit-code'.
```

I have manually created a home directory for the user, assigned the proper permissions and then restarted the service. After doing so I was able to launch a Python 3 notebook.

**Comment by yuvipanda** [comment]
Thanks for opening this, @owah! Can you provide logs from JupyterHub itself (`sudo journalctl -u jupyterhub`)? That should hopefully provide output on why the initial adduser command is failing.

**Comment by owah** [comment]
hi @yuvipanda, this is the weird thing, the command does not fail! I've checked the logs and all I see is:
```
Nov 23 21:50:11 juphubtest useradd[3081]: new group: name=jupyter-testuser, GID=1004
Nov 23 21:50:11 juphubtest useradd[3081]: new user: name=jupyter-testuser, UID=1002, GID=1004, home=/home/jupyter-testuser, shell=/bin/sh
```

**Comment by yuvipanda** [comment]
@owah interesting. is this repeatable? I wonder if adduser behavior changed in 18.l0.

**Comment by owah** [comment]
@yuvipanda it is.

What I have tried so far:
replace `subprocess.check_call()` with `os.system("useradd --create-home " + username)` and `os.system("useradd --create-home hardcoded-name")`

In both cases, the user gets added to `/etc/passwd` but no home-dir. is created.

I've also tried to make the home directory to be writeable for everybody (`ugo+rwx`) to see what happens. Nothing.

**Comment by owah** [comment]
I have just spun up a new VM, with Ubuntu Server 18.04.01 LTS and I can confirm, that there the user's home directory gets created without any issues.

I will investigate if/how adduser behavior has changed between these two releases.

**Comment by justinmoon** [comment]
@yuvipanda I've hit this bug 2 times in a row deploying 18.10 to digital ocean.

Logs from running `sudo journalctl -u jupyter` (I tried manually restarting the server from UI which is probably why you see "Adding user" twice ...):

```
Feb 05 05:32:41 jupyter systemd[1]: Started jupyterhub.service.
Feb 05 05:32:44 jupyter python3[8253]: [I 2019-02-05 05:32:44.220 JupyterHub app:1673] Using Authenticator: firstuseauthenticator.firstuseauthenticator.FirstUs
Feb 05 05:32:44 jupyter python3[8253]: [I 2019-02-05 05:32:44.220 JupyterHub app:1673] Using Spawner: builtins.UserCreatingSpawner
Feb 05 05:32:44 jupyter python3[8253]: [I 2019-02-05 05:32:44.223 JupyterHub app:1055] Writing cookie_secret to /opt/tljh/state/jupyterhub_cookie_secret
Feb 05 05:32:44 jupyter python3[8253]: [I 2019-02-05 05:32:44.272 alembic.runtime.migration migration:130] Context impl SQLiteImpl.
Feb 05 05:32:44 jupyter python3[8253]: [I 2019-02-05 05:32:44.273 alembic.runtime.migration migration:137] Will assume non-transactional DDL.
Feb 05 05:32:44 jupyter python3[8253]: [I 2019-02-05 05:32:44.283 alembic.runtime.migration migration:356] Running stamp_revision  -> 896818069c98
Feb 05 05:33:34 jupyter useradd[8411]: new group: name=jupyter-justin, GID=1002
Feb 05 05:33:34 jupyter useradd[8411]: new user: name=jupyter-justin, UID=1000, GID=1002, home=/home/jupyter-justin, shell=/bin/sh
Feb 05 05:33:34 jupyter python3[8253]: Adding user jupyter-justin to group jupyterhub-users
Feb 05 05:33:34 jupyter gpasswd[8418]: user jupyter-justin added by root to group jupyterhub-users
Feb 05 05:33:34 jupyter python3[8253]: Adding user jupyter-justin to group jupyterhub-admins
Feb 05 05:33:34 jupyter gpasswd[8422]: user jupyter-justin added by root to group jupyterhub-admins
Feb 05 05:33:34 jupyter python3[8253]: Running as unit: jupyter-justin.service
Feb 05 05:39:00 jupyter python3[8253]: Failed to start transient service unit: Unit jupyter-justin.service already exists.
```
Logs from `sudo journalctl -u jupyter-justin`:

```
-- Logs begin at Tue 2019-02-05 05:26:37 UTC, end at Tue 2019-02-05 05:43:21 UTC. --
Feb 05 05:33:34 jupyter systemd[1]: Started /bin/bash -c cd /home/jupyter-justin && exec jupyterhub-singleuser --port=36863.
Feb 05 05:33:34 jupyter systemd[8428]: jupyter-justin.service: Changing to the requested working directory failed: No such file or directory
Feb 05 05:33:34 jupyter systemd[8428]: jupyter-justin.service: Failed at step CHDIR spawning /bin/bash: No such file or directory
Feb 05 05:33:34 jupyter systemd[1]: jupyter-justin.service: Main process exited, code=exited, status=200/CHDIR
Feb 05 05:33:34 jupyter systemd[1]: jupyter-justin.service: Failed with result 'exit-code'.
```

**Comment by mkempenaar** [comment]
I am experiencing the same problem as well. However it seems to be coupled to the admin state of the user. Adding a new user without admin rights, logging out and logging in as the new user works fine. A new folder `/home/jupyter-user` is created and everything works as it should. 

Logging out the new user and trying to login the admin user results in the following error (indeed, the service file still exists):
`Failed to start transient service unit: Unit jupyter-user.service already exists.`. Manually (re-)starting this service results in the same permission denied error.

Manually creating the home directory for the admin user works though but this is weird behaviour.

**Comment by Alessandro-QDT** [comment]
Hi community. Did anyone found a solution to this issue?
It is still happening

**Comment by willirath** [comment]
No. But this really seems to be an issue with the firstuseauthenticator?

**Comment by johnmharris** [comment]
`/etc/systemd/system/jupyterhub.service` it defines:

```
# jupyterhub process should have no access to home directories
ProtectHome=tmpfs
```

Comment that out. The wisdom of that setting and the requirement for users to be created will need to be resolved by developers with more experience of jupyter than I though!

**Comment by EvilMav** [comment]
FYI: Issue persists with Kubuntu 19.04 - the home folder was not created at the first admin login.

**Comment by johnmharris** [comment]
> FYI: Issue persists with Kubuntu 19.04 - the home folder was not created at the first admin login.

Did you try disabling ProtectHome=tmpfs ?

**Comment by EvilMav** [comment]
@johnmharris yes, creating new users did work after this. I had to create first user's home directory manually for it to work afterwards.

**Comment by johnmharris** [comment]
Without ProtectHome the user directories should be created when useradd is invoked, as @owah noted.

**Comment by minrk** [comment]
closed by #435
